File: frontend/src/types/transportadora.ts
export interface Transportadora {
  id: string;
  codigo?: string | null;
  nome: string;
  razao_social: string;
  cnpj_cpf: string | null;
  segmento: string;
  tipo_integracao: TipoIntegracao;
  metodo_calculo: MetodoCalculo;
  api_ambiente: "producao" | "homologacao" | null;
  status_integracao: "pendente_credencial" | "ativo" | "erro" | "nao_aplicavel";
  ativa: boolean;
  taxa_sucesso: number;
  tempo_medio_ms: number;
  precisa_revisao?: boolean;
  status_validacao?: string;
  cidade?: string | null;
  uf?: string | null;
  nome_fantasia?: string | null;
  rntrc?: string | null;
  site?: string | null;
  email_comercial?: string | null;
  telefone?: string | null;
  cep?: string | null;
}

export type TipoIntegracao = "api" | "tabela" | "webservice" | "soap" | "edi" | "n8n" | "playwright";
export type MetodoCalculo = "tabela_propria" | "api" | "webservice" | "manual";

export interface TransportadoraInput {
  nome: string;
  razao_social: string;
  cnpj_cpf: string;
  segmento: string;
  tipo_integracao: TipoIntegracao;
  metodo_calculo?: MetodoCalculo;
  api_base_url?: string | null;
  api_key?: string | null;
  api_ambiente?: "producao" | "homologacao" | null;
  ativa?: boolean;
}

export interface ConsultaCnpj {
  cnpj: string;
  nome_fantasia: string;
  razao_social: string;
  segmento: string | null;
  situacao_cadastral: string | null;
  cep: string | null;
  municipio: string | null;
  uf: string | null;
}

export interface ConfiguracaoApi {
  transportadora_id: string;
  base_url: string;
  endpoint_cotacao: string;
  metodo_http: "GET" | "POST";
  tipo_autenticacao: "bearer" | "api_key" | "basic" | "nenhuma" | "jamef_login" | "braspress_basic";
  nome_header: string | null;
  usuario_integracao?: string | null;
  auth_url?: string | null;
  documento_devedor?: string | null;
  filial_origem?: string | null;
  tipo_transporte?: "1" | "2" | "R" | "A" | null;
  campo_valor: string;
  campo_prazo: string;
  ativa: boolean;
  credencial_configurada: boolean;
  credencial_mascarada: string | null;
}

export interface ConfiguracaoApiInput {
  base_url: string;
  endpoint_cotacao: string;
  metodo_http: "GET" | "POST";
  tipo_autenticacao: "bearer" | "api_key" | "basic" | "nenhuma" | "jamef_login" | "braspress_basic";
  nome_header?: string | null;
  credencial?: string | null;
  usuario_integracao?: string | null;
  auth_url?: string | null;
  documento_devedor?: string | null;
  filial_origem?: string | null;
  tipo_transporte?: "1" | "2" | "R" | "A" | null;
  campo_valor: string;
  campo_prazo: string;
  ativa: boolean;
}

export interface MapeamentoSankhya {
  id: string;
  transportadora_id: string;
  codigo_parceiro: number;
  nome_parceiro: string;
  codigo_servico: string | null;
  servico: string | null;
  ativo: boolean;
}

export type MapeamentoSankhyaInput = Omit<MapeamentoSankhya, "id">;

export type ResultadoImportacao = "NOVO" | "ATUALIZACAO" | "IGNORADO" | "REVISAO" | "ERRO" | "DUPLICADO";
export interface ImportacaoItem { linha: number; codigo_importacao: string | null; nome_transportadora: string | null; cnpj: string | null; resultado: ResultadoImportacao; avisos: string[]; erros: string[]; dados_normalizados: Record<string, unknown>; }
export interface ImportacaoPreview { import_id: string; total: number; novos: number; atualizacoes: number; ignorados: number; revisao: number; erros: number; registros: ImportacaoItem[]; }
export interface ImportacaoResultado { import_id: string; status: string; resultado: Record<string, number>; transportadora_ids: string[]; }

export type CarrierIntegrationType = "API" | "TABLE" | "HYBRID" | "MANUAL" | "RPA";
export interface CarrierService { id: string; carrier_id: string; name: string; code: string; external_code: string | null; description: string | null; service_type: string | null; active: boolean; }
export interface CarrierIntegration { id: string; carrier_id: string; integration_type: CarrierIntegrationType; adapter_code: string | null; active: boolean; priority: number; configuration: Record<string, unknown>; status: "not_configured" | "configured" | "validated" | "error" | "inactive"; credential_keys: string[]; }
export interface CredentialStatus { configured: boolean; keys: string[]; masked: Record<string, string>; }

export interface SSWIntegration {
  transportadora_id: string;
  provider: "SSW";
  dominio: string;
  login: string;
  cnpj_pagador: string;
  mercadoria_padrao: number;
  ativo: boolean;
  credencial_configurada: boolean;
  ultimo_teste: string | null;
  status_ultima_validacao: "NAO_TESTADA" | "VALIDA" | "INVALIDA";
  mensagem_ultima_validacao: string | null;
}

export interface AnttTransportadora {
  nome: string;
  cnpj: string;
  rntrc: string;
  situacao: string;
  categoria: "ETC" | "CTC";
  municipio: string | null;
  uf: string | null;
  cep: string | null;
  ja_cadastrada: boolean;
}

export interface SSWIntegrationInput {
  dominio: string;
  login: string;
  senha?: string | null;
  cnpj_pagador: string;
  mercadoria_padrao: number;
  ativo: boolean;
}

export interface EnrichmentStatus { transportadora_id:string; status:string; enrichment_started_at:string|null; enrichment_finished_at:string|null; last_enrichment_at:string|null; completion_percent:number; counts:Record<string,number>; }
export interface EnrichmentCoverage { id:string; coverage_type:string; uf:string|null; city:string|null; cep_start:string|null; cep_end:string|null; pickup_available:boolean; delivery_available:boolean; confidence_score:number; source_url:string; }
export interface EnrichmentIntegration { id:string; integration_type:string; provider:string|null; status:string; url:string|null; documentation_url:string|null; is_public:boolean; confidence_score:number|null; source_url:string|null; }
export interface EnrichmentBranch { id:string; name:string; cnpj:string|null; cep:string|null; city:string|null; uf:string|null; phone:string|null; email:string|null; confidence_score:number; source_url:string; }
export interface EnrichmentSource { id:string; tipo_fonte:string; url:string|null; http_status:number|null; confidence_score:number|null; data_pesquisa:string|null; }
export interface EnrichmentEvidence { id:string; evidence_type:string; value:string; evidence:Record<string,unknown>; confidence_score:number; review_status:"PENDING"|"APPROVED"|"REJECTED"; source_url:string; }
export interface CarrierSummary { transportadora_id:string; status:string; completion_percent:number; last_enrichment_at:string|null; integration_types:string[]; coverage_states:string[]; national_coverage:boolean; branches_count:number; freight_table_access:string|null; pending_reviews:number; confidence_score:number|null; next_verification_at:string|null; }
export interface CarrierEligibility { transportadora_id:string; transportadora:string; origin_coverage:boolean; destination_coverage:boolean; integration_type:string; confidence_score:number; eligible:boolean; }
export interface CarrierCardIntegration { type:string; detected:boolean; configured:boolean; active:boolean; }
export interface CarrierCard { id:string; nome:string; razao_social:string; cnpj:string|null; rntrc:string|null; ativa:boolean; integrations:CarrierCardIntegration[]; coverage:{identified:boolean;national:boolean;states:string[];total_states:number}; enrichment:{status:string;percentage:number;last_run_at:string|null}; branches_count:number; has_freight_table:boolean; freight_table_access:string|null; }
export interface CarrierCardsPage { items:CarrierCard[]; page:number; page_size:number; total:number; pages:number; }
export interface CarrierStats { total:number; active:number; with_api:number; with_ssw:number; with_table:number; pending_enrichment:number; }
export interface EnrichmentJob { id:string; transportadora_id:string; status:"QUEUED"|"PROCESSING"|"SUCCESS"|"FAILED"|"CANCELLED"; progress:number; current_step:string|null; started_at:string|null; finished_at:string|null; error_message:string|null; }
export interface CoverageCheck { pickup:boolean; delivery:boolean; eligible:boolean; }

File: backend/app/schemas/transportadora.py
import re
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, field_validator


TIPOS_INTEGRACAO = {"api", "tabela", "webservice", "soap", "edi", "n8n", "playwright"}
METODOS_CALCULO = {"tabela_propria", "api", "webservice", "manual"}


def somente_digitos(valor: str) -> str:
    return re.sub(r"\D", "", valor)


def documento_valido(valor: str) -> bool:
    digitos = somente_digitos(valor)
    if len(digitos) not in {11, 14} or len(set(digitos)) == 1:
        return False

    if len(digitos) == 11:
        base = [int(numero) for numero in digitos[:9]]
        primeiro = (sum(numero * peso for numero, peso in zip(base, range(10, 1, -1))) * 10) % 11
        primeiro = 0 if primeiro == 10 else primeiro
        segundo = (sum(numero * peso for numero, peso in zip(base + [primeiro], range(11, 1, -1))) * 10) % 11
        segundo = 0 if segundo == 10 else segundo
        return digitos[-2:] == f"{primeiro}{segundo}"

    def digito_cnpj(base: list[int], pesos: list[int]) -> int:
        resto = sum(numero * peso for numero, peso in zip(base, pesos)) % 11
        return 0 if resto < 2 else 11 - resto

    base_cnpj = [int(numero) for numero in digitos[:12]]
    primeiro = digito_cnpj(base_cnpj, [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2])
    segundo = digito_cnpj(base_cnpj + [primeiro], [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2])
    return digitos[-2:] == f"{primeiro}{segundo}"


class TransportadoraBase(BaseModel):
    nome: str = Field(min_length=2, max_length=120)
    razao_social: str = Field(min_length=2, max_length=255)
    cnpj_cpf: str | None
    segmento: str = Field(min_length=2, max_length=80)
    tipo_integracao: str
    metodo_calculo: str | None = None
    api_ambiente: str | None = None

    @field_validator("nome", "razao_social", "segmento")
    @classmethod
    def limpar_texto(cls, valor: str) -> str:
        return valor.strip()

    @field_validator("cnpj_cpf")
    @classmethod
    def validar_documento(cls, valor: str | None) -> str | None:
        if valor is None:
            return None
        normalizado = somente_digitos(valor)
        if not documento_valido(normalizado):
            raise ValueError("CPF ou CNPJ inválido")
        return normalizado

    @field_validator("tipo_integracao")
    @classmethod
    def validar_tipo_integracao(cls, valor: str) -> str:
        normalizado = valor.strip().lower()
        if normalizado not in TIPOS_INTEGRACAO:
            raise ValueError("Tipo de integração inválido")
        return normalizado

    @field_validator("metodo_calculo")
    @classmethod
    def validar_metodo_calculo(cls, valor: str | None) -> str | None:
        if valor is None:
            return None
        normalizado = valor.strip().lower()
        if normalizado not in METODOS_CALCULO:
            raise ValueError("Método de cálculo inválido")
        return normalizado


class TransportadoraCreate(TransportadoraBase):
    ativa: bool = True
    api_key: str | None = Field(None, max_length=4000)
    api_base_url: HttpUrl | None = None


class TransportadoraUpdate(BaseModel):
    nome: str | None = Field(None, min_length=2, max_length=120)
    razao_social: str | None = Field(None, min_length=2, max_length=255)
    cnpj_cpf: str | None = None
    segmento: str | None = Field(None, min_length=2, max_length=80)
    tipo_integracao: str | None = None
    metodo_calculo: str | None = None
    api_ambiente: str | None = None
    api_key: str | None = Field(None, max_length=4000)
    api_base_url: HttpUrl | None = None

    @field_validator("nome", "razao_social", "segmento")
    @classmethod
    def limpar_texto(cls, valor: str | None) -> str | None:
        return valor.strip() if valor is not None else None

    @field_validator("cnpj_cpf")
    @classmethod
    def validar_documento(cls, valor: str | None) -> str | None:
        if valor is None:
            return None
        normalizado = somente_digitos(valor)
        if not documento_valido(normalizado):
            raise ValueError("CPF ou CNPJ inválido")
        return normalizado

    @field_validator("tipo_integracao")
    @classmethod
    def validar_tipo_integracao(cls, valor: str | None) -> str | None:
        if valor is None:
            return None
        normalizado = valor.strip().lower()
        if normalizado not in TIPOS_INTEGRACAO:
            raise ValueError("Tipo de integração inválido")
        return normalizado

    @field_validator("metodo_calculo")
    @classmethod
    def validar_metodo_calculo(cls, valor: str | None) -> str | None:
        if valor is None:
            return None
        normalizado = valor.strip().lower()
        if normalizado not in METODOS_CALCULO:
            raise ValueError("Método de cálculo inválido")
        return normalizado


class TransportadoraStatusUpdate(BaseModel):
    ativa: bool


class ConsultaCnpjOut(BaseModel):
    cnpj: str
    nome_fantasia: str
    razao_social: str
    segmento: str | None = None
    situacao_cadastral: str | None = None
    cep: str | None = None
    municipio: str | None = None
    uf: str | None = None


class TransportadoraOut(TransportadoraBase):
    model_config = ConfigDict(from_attributes=True)

    id: str
    ativa: bool
    taxa_sucesso: float
    tempo_medio_ms: int
    status_integracao: str
    codigo_importacao: str | None = None
    status_cnpj: str | None = None
    integracao_disponivel: str | None = None
    site: str | None = None
    portal_cotacao: str | None = None
    api_documentacao: str | None = None
    email_comercial: str | None = None
    telefone: str | None = None
    cidade: str | None = None
    uf: str | None = None
    cep: str | None = None
    precisa_revisao: bool = False
    status_validacao: str = "A_VALIDAR"


class ConfiguracaoApiUpdate(BaseModel):
    base_url: HttpUrl
    endpoint_cotacao: str = Field(default="", max_length=500)
    metodo_http: str = "POST"
    tipo_autenticacao: str = "bearer"
    nome_header: str | None = Field(None, max_length=120)
    credencial: str | None = Field(None, max_length=4000)
    usuario_integracao: str | None = Field(None, max_length=255)
    auth_url: HttpUrl | None = None
    documento_devedor: str | None = Field(None, min_length=11, max_length=14)
    filial_origem: str | None = Field(None, max_length=10)
    tipo_transporte: Literal["1", "2"] | None = None
    campo_valor: str = Field(default="valor_frete", min_length=1, max_length=200)
    campo_prazo: str = Field(default="prazo_dias", min_length=1, max_length=200)
    ativa: bool = False

    @field_validator("metodo_http")
    @classmethod
    def validar_metodo(cls, valor: str) -> str:
        normalizado = valor.upper()
        if normalizado not in {"GET", "POST"}:
            raise ValueError("Método deve ser GET ou POST")
        return normalizado

    @field_validator("tipo_autenticacao")
    @classmethod
    def validar_autenticacao(cls, valor: str) -> str:
        normalizado = valor.lower()
        if normalizado not in {"bearer", "api_key", "basic", "nenhuma", "jamef_login"}:
            raise ValueError("Tipo de autenticação inválido")
        return normalizado


class ConfiguracaoApiOut(BaseModel):
    transportadora_id: str
    base_url: str
    endpoint_cotacao: str
    metodo_http: str
    tipo_autenticacao: str
    nome_header: str | None
    usuario_integracao: str | None = None
    auth_url: str | None = None
    documento_devedor: str | None = None
    filial_origem: str | None = None
    tipo_transporte: str | None = None
    campo_valor: str
    campo_prazo: str
    ativa: bool
    credencial_configurada: bool
    credencial_mascarada: str | None = None


class CredencialUpdate(BaseModel):
    credencial: str = Field(min_length=1, max_length=4000)


class StatusIntegracaoOut(BaseModel):
    transportadora_id: str
    metodo_calculo: str
    status_integracao: str
    pronta_para_cotacao: bool
    mensagem: str


class ImportacaoItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    linha: int
    codigo_importacao: str | None
    nome_transportadora: str | None
    cnpj: str | None
    resultado: str
    avisos: list[str]
    erros: list[str]
    dados_normalizados: dict


class ImportacaoPreviewOut(BaseModel):
    import_id: str
    total: int
    novos: int
    atualizacoes: int
    ignorados: int
    revisao: int
    erros: int
    registros: list[ImportacaoItemOut]


class ImportacaoConfirmIn(BaseModel):
    atualizar_existentes: bool = True
    importar_em_revisao: bool = False


class ImportacaoResultadoOut(BaseModel):
    import_id: str
    status: str
    resultado: dict[str, int]
